Add support for constant_pad_nd

ljfitz opened the pull request:
Note that to enable folding of the code coming from an example
like the ConstantPad2dStaticModule e2e test, support for other
operations had to be added/improved:
- aten::neg.int
- aten::eq.float
- aten::eq.str
- prim::Uninitialized

cathyzhyi commented on e2e_testing/torchscript/basic.py:
```diff
@@ -284,6 +284,47 @@ def MaxPool2dModule_basic(module, tu: TestUtils):
+def ConstantPad2dStaticModule_basic(module, tu: TestUtils):
+    module.forward(tu.rand(1, 1, 20, 20) - 0.5)
+
+
```
We usually put the delimiter before the testing module here. Can you also help to fix the existing one for `MaxPool2dModule_basic`

cathyzhyi commented on include/torch-mlir/Dialect/Torch/IR/TorchOps.h:
```diff
@@ -45,6 +45,21 @@ struct torch_constant_int_op_binder {
 };
+
+struct torch_constant_float_op_binder {
+  APFloat *bind_value;
```
Prefer to use the primitive type `double` so that user of the API can use the result number directly for calculation. Like in `AtenEqFloatOp::fold`, `lhs.compare(rhs) == APFloat::cmpEqual` can be replaced with `lhs == rhs`.

cathyzhyi commented on lib/Dialect/Torch/IR/TorchOps.cpp:
```diff
@@ -1005,6 +1034,20 @@ void PrimTupleIndexOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+void PrimUninitializedOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                                                    MLIRContext *context) {
+  patterns.add(+[](PrimUninitializedOp op, PatternRewriter &rewriter) {
+    if (! op.use_empty())
```
We shouldn't need to add such patterns because the canonicalizer should remove used ops without side effects automatically. I think the reason this is not removed is that the op might have side affects. 

Is this op on some dead code path?

ljfitz replied:
For the e2e example, without this a `"prim.Uninitialized"` is present unconditionally but the result is unused. I cannot track down a side-effect that would require it to remain present. Any hints as to where it make be treated as having a (pseudo) side-effect that prevents it being otherwise removed?

cathyzhyi replied:
`side-effect` doesn't mean there must be uses. It could just be an op that can through exception or can allocate memory. If an op doesn't have side effect, the op definition need to contain the `NoSideEffect` trait like this one https://github.com/llvm/mlir-npcomp/blob/d6b6c0268cfd6ead0aa5f2f21ba02bbe8ead3180/include/torch-mlir/Dialect/Torch/IR/TorchOps.td#L798.  Currently, we don't support generating that trait for auto generated ops ods yet. 

However, putting all these aside,  according to https://discuss.pytorch.org/t/keyerror-prim-uninitialized/64860 `prim.Uninitialized` means the compiler believe the code can never reach this statement. It's very weird to see the `prim.Uninitialzied` on an unconditional path.

cathyzhyi replied:
Oh, actually, it didn't say it's on dead code path and but the value is not used. Hmm, in this case we could just delete the op unconditionally for now.

ljfitz replied:
I'd seen that discussion and considered trying to remove it earlier than the point here. However in the incoming IR the result SSA is referenced, but after some folding and simplifications the unused `prim.Uninitialized` remains. This was the first point I could identify that it was straightforward to remove it. I'm happy to dig further, but would appreciate some pointers if you have them.

cathyzhyi replied:
I see. yea, we can leave it as it is then. btw, did you use `clang-format`? seems there is a space after `!`

ljfitz replied:
I'd run it manually but it looks like I forgot to do so after a change. Do you use https://pre-commit.com/ or similar to avoid forgetting?

cathyzhyi commented on lib/Dialect/Torch/IR/TorchOps.cpp:
```diff
@@ -1129,6 +1172,18 @@ OpFoldResult AtenMulIntOp::fold(ArrayRef<Attribute> operands) {
+OpFoldResult AtenNegIntOp::fold(ArrayRef<Attribute> operands) {
+  int64_t c;
+  bool constant = matchPattern(getOperand(), m_TorchConstantInt(&c));
+  if (constant)
```
nit: don't need the `constant` variable.

cathyzhyi commented on lib/Dialect/Torch/IR/TorchOps.cpp:
```diff
@@ -653,6 +655,33 @@ OpFoldResult AtenEqIntOp::fold(ArrayRef<Attribute> operands) {
+// AtenEqFloatOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult AtenEqFloatOp::fold(ArrayRef<Attribute> operands) {
```
all folders need unit tests in https://github.com/llvm/mlir-npcomp/blob/d6b6c0268cfd6ead0aa5f2f21ba02bbe8ead3180/test/Dialect/Torch/canonicalize.mlir

cathyzhyi reviewed (changes requested):
I did a first pass. Will take a closer look at torch lowering and refine types later.

cathyzhyi commented on lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:
```diff
@@ -2685,6 +2696,58 @@ class ConvertAtenMaxPool2dOp : public OpConversionPattern<AtenMaxPool2dOp> {
+    }
+
+    Value castedValue =
+        rewriter.create<arith::TruncFOp>(loc, type.getElementType(), adaptor.value());
```
the `convertScalarToDtype` helper can be used to covert the `value` type to tensor element type. 

Also, technically, the `elementType` need to be got from the converted result type, `getTypeConverter()->convertType(op.getType())`;.  This is not an issue for float because the element type of converted tensor type and the torch tensor type are the same. However, it would be different for integer tensor types.

ljfitz replied:
Thanks for the explanation!

cathyzhyi commented on lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:
```diff
@@ -2685,6 +2696,58 @@ class ConvertAtenMaxPool2dOp : public OpConversionPattern<AtenMaxPool2dOp> {
+    int64_t rank = type.getRank();
+
+    Type elementType = self.getType().cast<RankedTensorType>().getElementType();
+    if (!elementType.isa<mlir::FloatType>())
```
This type limitation can be removed once the `arith::TruncFOp` below is changed to use `convertScalarToDtype`

cathyzhyi commented on lib/Dialect/Torch/Transforms/RefineTypes.cpp:
```diff
@@ -953,6 +958,28 @@ ChangeResult TypeAnalyzer::visitAtenMaxPool2dOp(
+  auto knowledge =
+      ValueKnowledge::getNotNonePessimisticValueState(op->getContext());
+  knowledge.hasSizes = true;
+  auto &ifm = operands[0]->getValue();
```
we don't need the & as the return type of `getValue` is already reference type.

cathyzhyi replied:
nit: not sure why is the variable name `ifm`? Can we call it `input`?

ljfitz replied:
> we don't need the & as the return type of `getValue` is already reference type

While that holds true for `auto` and pointer types, the `&` is necessary if you want a reference rather than a copy (e.g. 
https://github.com/llvm/torch-mlir/pull/515#discussion_r781478778).

cathyzhyi replied:
ha, you are right.

cathyzhyi commented on lib/Dialect/Torch/Transforms/RefineTypes.cpp:
```diff
@@ -953,6 +958,28 @@ ChangeResult TypeAnalyzer::visitAtenMaxPool2dOp(
+    knowledge.sizes = ifm.sizes;
+    uint64_t padRank = padInts.size() / 2;
+    uint64_t padOffset = knowledge.sizes.size() - padRank;
+    // op.pad() is higest dim first ordered pairs of low,high
```
nit: period at the end of comment.

cathyzhyi commented on lib/Dialect/Torch/Transforms/RefineTypes.cpp:
```diff
@@ -953,6 +958,28 @@ ChangeResult TypeAnalyzer::visitAtenMaxPool2dOp(
+      knowledge.sizes[r] += padInts[i * 2 - 2] + padInts[i * 2 - 1];
+    }
+  } else
+    knowledge.sizes.resize(4, kUnknownSize);
```
we don't need the else branch. The `knowledge` is initialized with unknown size. If you would like to be more accurate. You could arrange the branches like
```
if (input.hasSizes) {
   knowledge.hasSizes = true;
   if (matchPattern(op.pad(), m_TorchConstantIntList(padInts))){
     fill in sizes here....
   } else {
     knowledge.sizes.resize(inputRank, kUnknownSize);
   }
}
```

cathyzhyi commented on lib/Dialect/Torch/Transforms/RefineTypes.cpp:
```diff
@@ -953,6 +958,28 @@ ChangeResult TypeAnalyzer::visitAtenMaxPool2dOp(
+    ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto knowledge =
+      ValueKnowledge::getNotNonePessimisticValueState(op->getContext());
+  knowledge.hasSizes = true;
```
this need to be deleted at this line as we might not have sizes information for the output.

cathyzhyi commented on e2e_testing/torchscript/basic.py:
```diff
@@ -284,6 +284,47 @@ def MaxPool2dModule_basic(module, tu: TestUtils):
+    @export
+    @annotate_args([
+        None,
+        ([1, 1, 20, 20], torch.float32, True),
```
Can we also add testing for dynamic shapes?

cathyzhyi commented on lib/Dialect/Torch/Transforms/RefineTypes.cpp:
```diff
@@ -953,6 +958,28 @@ ChangeResult TypeAnalyzer::visitAtenMaxPool2dOp(
+    uint64_t padOffset = knowledge.sizes.size() - padRank;
+    // op.pad() is higest dim first ordered pairs of low,high
+    for (uint64_t i = padRank, r = padOffset; i > 0; --i, ++r) {
+      knowledge.sizes[r] += padInts[i * 2 - 2] + padInts[i * 2 - 1];
```
need to guard the calculation with if `knowledge.sizes[r] == kunknownSize`

ljfitz commented:
Thanks @cathyzhyi for the careful review - very much appreciated. I think I've addressed everything, though I'm not sure if you find the handling of `prim.Uninitialized` acceptable for this PR or not.

As there were quite a few comments I've pushed a separate commit for now to make it easier for you to check. I'll rebase and squash when the PR is approved.

cathyzhyi reviewed (approved):
LGTM. Thanks for the PR!

ljfitz commented:
> LGTM. Thanks for the PR!

And thank you for the thorough and supportive review. I've rebased, clang-formatted and squashed the commits. I see that I haven't permission to merge the the pull request. Do you know if merging will happen automatically or if I need to take some additional action?
